Static GUE decapsulation README (#4178)
openconfig/featureprofiles: feature/policy_forwarding/encapsulation/otg_tests/decap_static_guev1/README.md

Author: Done. I have moved it with other decap folders `feature/policy_forwarding/otg_tests/ip_guev1_static_decap/README.md`

# PF-1.4: Static GUEv1 Decapsulation to IPv4 or IPv6 payload

## Summary

> This is to test the functionality of decapsulation of static GUEv1 to IPv4 or IPv6 payload ([GUE Variant 1](https://datatracker.ietf.org/doc/html/draft-ietf-intarea-gue-09#section-4)). These tests verify the use case of IPv4 and IPv6 encapsulated traffic in IPv4 GUE tuennel. 

Reviewer: ```suggestion
This is to test the functionality of decapsulation of static GUEv1 to IPv4 or IPv6 payload ([GUE Variant 1](https://datatracker.ietf.org/doc/html/draft-ietf-intarea-gue-09#section-4)). These tests verify the use case of IPv4 and IPv6 encapsulated traffic in IPv4 GUE tunnel. 
```

The tests validate that the DUT performs the following action-

 - Decapsulate the outer ((UDPoIPv4)) headers of GUE packets destined to the locally configured decap IPv4 address/addresses and matching UDP port. 

>     - The first two bits the UDP payload will be used to determine GUE varient 1 (MSB bits 01) and first four bits will determine the payload protocol IPv4/IPv6.

Reviewer: ```suggestion
    - The first two bits of the UDP payload will be used to determine GUE varient 1 (MSB bits 01) and first four bits will determine the payload protocol version IPv4/IPv6.
```

-  The outer header destination port of the traffic (UDP 6081) matches the configured non-default UDP decap port criteria
-  The DUT will decapsulate the outer header and perform the lookup based on the inner IP address
-  The TTL bits will not be copied to the inner header
-  The inner header TTL value will be decremented by 1
-  The inner packet's DSCP value (32) will be preserved
-  The DUT will forward the traffic towards ATE Port 2
-  The relevant DUT counters will reflect 1,000,000 decapsulated packets
-  ATE Port 2 receives 1,000,000 packets structured as RX-Flow-type#4
-  No packet loss should be observed

### PF-1.4.5: GUE Decapsulation of inner IPv4 traffic using non-default and unconfigured GUE UDP port (Negative).
-  Push DUT configuration.
-  Initiate traffic flow type#5
Verification: 
-  The outer header destination IP of the traffic will fall within the decap range configured(DECAP-DST-SUBNET/30)
-  The outer header destination port of the traffic is unconfigured UDP decap port, therefore it doesn't matches the criteria and the relevant packets should be dropped

> -  The DUT will decapsulate the outer header and perform the lookup based on the inner ip address

Reviewer: if the UDP port doesnt match, how would the decap happen? Also, this seems counter to the bullet above?

Author: Updated

        }
    }
}
```

## OpenConfig Path and RPC Coverage

```yaml

# paths:

/network-instances/network-instance/policy-forwarding/policies/policy/config/policy-id:
/network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/config/ipv4/config/destination-address-prefix-set:
/network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/config/ipv4/config/protocol:
/network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/transport/config/destination-port:
/network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/action/decapsulate-gue:

# telemetry
openconfig-interfaces/interfaces/interface/state/counters/out-pkts:
openconfig-interfaces/interfaces/interface/state/counters/out-unicast-pkts:
openconfig-network-instance/network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/state/matched-pkts:
openconfig-network-instance/network-instances/network-instance/policy-forwarding/policies/policy/rules/rule/state/matched-octets:

rpcs:
  gnmi:
    gNMI.Set:
      union_replace: true
      replace: true
    gNMI.Subscribe:
      on_change: true
```

## Required DUT platform

> * Specify the minimum DUT-type:

Reviewer: minimum can just be vRX?

Reviewer 2: IMO,  decap is very sensitive to the ASIC implementation, (this isn't a control-plane only test) so this should be done on FFF.

Author: ack. done

# PF-1.4: Static GUEv1 Decapsulation to IPv4 or IPv6 payload

## Summary

This is to test the functionality of decapsulation of static GUEv1 to IPv4 or IPv6 payload ([GUE Variant 1](https://datatracker.ietf.org/doc/html/draft-ietf-intarea-gue-09#section-4)). These tests verify the use case of IPv4 and IPv6 encapsulated traffic in IPv4 GUE tunnel. 

The tests validate that the DUT performs the following action-

 - Decapsulate the outer (UDPoIPv4) headers of GUE packets destined to the locally configured decap IPv4 address/addresses and matching UDP port. 
    - The first two bits of the UDP payload will be used to determine GUE varient 1 (MSB bits 01) and first four bits will determine the payload protocol version IPv4/IPv6.
    - If the first four bits of the payload protocol has bit combination of 0100, it specifies the inner payload protocol is IPv4.

>     - If the first four bits of the payload protocol has bit combination of 0110, it specifies the inner payload protocol is IPv6. 

Reviewer 2: This text can be dropped, it's already specified in the IETF draft linked above.

Author: done

 - Decapsulate the packet only if it matches the locally configured decap IPv4 address/addresses and matching UDP port port/port-range.

>     - If not specifically configured, then the DST UDP port 6080 will be used as a match criteria for the decapsulation.

Reviewer 2: The OC requires specific configuration, there is no default.  Any default behavior is undefined and does not need to be tested and cannot be assumed to be 6080.  

```suggestion
    - Configure decap policy matching on destination UDP port 6080.
```

An alternative to this configuration text is to simply refer to the canonical OC JSON which specifies the rules and actions.

Author: Updated this statement and aligned the test cases accordingly. (Updated as per RFC and OC support)

>  - Post decapsulation the DUT should not copy outer TTL to inner header and maintain the inner DSCP vaule as is.

Reviewer 2: ```suggestion
 - Post decapsulation the DUT should 
     - Maintain the inner TTL (do not copy outer TTL to the inner TTL).  
     - Maintain the inner DSCP (do not copy outer DSCP to inner DSCP).
```

Author: done

>     - If explicit configration is present to copy the TTL from outer to inner header after decapsulation, then it will be honored.

Reviewer 2: Currently the test is missing a subtest which verifies this behavior.  If we don't require this functionality, then this statement can be left out.   The specification for how TTL configuration is supposed to work is covered in the  openconfig YANG data model at: 

https://github.com/openconfig/public/blob/785b890a6369901d5cc227f1a86de98b694305d7/release/models/aft/openconfig-aft-common.yang#L648-L655

Any clarifications to the specification should be done in the YANG file, not here in featureprofiles.

Author: Removed the statement

>  - Post decapsulation, the inner TTL value will be decremented before egressing to next-hop

Reviewer 2: Isn't this supposed to happen on ingress?  In the case of TTL=1,  why encap a packet and send if across the network onnly to be dropped on the decap end?

Author: The [RFC](https://screenshot.googleplex.com/6VnA6x6ESy3F6iA) says, IP-in-UDP encap packet should follow RFC2003 TTL behavior.
As per [RFC2003](https://screenshot.googleplex.com/B4ySfa3hnsAFvQr) the encapsulator will decrement the TTL by 1 before encapsulating the packet and the decapsulator will decrement the packet by 1 after decapsulation before sending out the packet to next-hop. 

On any node before egressing the packet if TTL = 0 after decrement then the node will drop the packet and not forward it to next-hop or destination. This is also applicable on encapsulator node if before encapsulation the TTL = 0 after decrement then the packet is drop locally.

    *  Configure the DUT as GUEv1 decapsulator with below fields:
       - DECAP-DST-SUBNET/30 as decapsulation subnet-range
          - DECAP-DST/32 is one of the IP prefix from DECAP-DST-SUBNET/30
       - Do not enable copying TTL bits from outer to inner header post decapsulation
       - Use single UDP destination port 6080 for both IPv4 and IPv6 payload type
    *  Configure static routes for destination IPV4-DST2 and IPV6-DST2 towards ATE Port 2
    *  Configure the UDP decap destination port as 6081 

* DECAP-DST-SUBNET/30 subnet comprises of below IP prefixes:
   - DECAP-DST_1/32
   - DECAP-DST_2/32
   - DECAP-DST_3/32
   - DECAP-DST_4/32

* DUT hosted IPv4 prefixes:
  * Loopback Address = DECAP-DST_3/32

### Test Procedure
Traffic: 
-  Generate the applicable GUE-encapsulated flow from ATE Port 1
     -  Sent 1000000 packets at the 10%  of the line rate.
     -  Keep the frame size Internet Mix.
   
### PF-1.4.1: GUE Decapsulation of inner IPv4 traffic using default GUE UDP port 6080
-  Push DUT configuration
-  Initiate traffic flow type#1  
Verification:
-  The outer header destination IP of the traffic falls within the configured decap range (DECAP-DST-SUBNET/30)

> -  The outer header destination port of the traffic (UDP 6080) matches the default UDP decap port criteria

Reviewer 2: as noted above, there is no default in the OC model for this, the configuration must specify the destination port used

Reviewer 2: nit:  this conflicts with line 97:
https://github.com/openconfig/featureprofiles/pull/4178/files#diff-cfed542cee25fba0c142d4357accaac8d378a13886f4eb6872a62b25177be9b6R97

-  The DUT will decapsulate the outer header and perform a lookup based on the inner IP address
-  The TTL bits will not be copied to the inner header
-  The inner header TTL value will be decremented by 1
-  The inner packet's DSCP value (32) will be preserved
-  The DUT will forward the traffic towards ATE Port 2

> -  The relevant DUT counters will reflect 1,000,000 decapsulated packets

Reviewer 2: Here and throughout, since we are not testing load balancing or performance here, why send 1M packets?  This makes the test take a longer time to execute.  Sending more than 1 packet seems like a good idea, but perhaps this could be reduced to just 1,000 packets?

Author: Updated

-  ATE Port 2 receives 1,000,000 packets structured as RX-Flow-type#1
-  No packet loss should be observed
  
### PF-1.4.2: GUE Decapsulation of inner IPv4 traffic using non-default and configured GUE UDP port

> -  Push DUT configuration

Reviewer 2: Specify what is different about the configuration here as compared to the canonical OC (provide the path)

Author: I don't follow your comment. This statement simply means pushing the OC config

### PF-1.4.8: Inner IPV6 GUE Pass-through (Negative)
-  Push DUT configuration.
-  Initiate traffic flow type#8
Verification:
-  The outer header destination UDP port (6081) matches a configured decap port
-  The outer header destination IP (ATE-P2-Address) of the traffic does not fall within the configured decap range (DECAP-DST-SUBNET/30), therefore it does not match the decapsulation criteria for the destination IP
-  The DUT will not decapsulate the outer header. Instead, it will perform a lookup based on the outer destination IP address and forward the packets as standard IP traffic
-  ATE Port 2 receives 1,000,000 packets structured as RX-Flow-type#8 (original GUE encapsulated packets, outer TTL decremented by DUT)
-  No packet loss should be observed

  ## Canonical OpenConfig for GUEv1 Decapsulation configuration

```json
{
    "network-instances": {
        "network-instance": {
            "config": {
                "name": "DEFAULT"
            },
            "name": "DEFAULT",
            "policy-forwarding": {
                "policies": {
                    "policy": [
                        {
                            "config": {

>                                 "policy-id": "decap-policy"

Author: Done
